### PR TITLE
Return complete history

### LINKED ISSUE #520: history() function returns an incomplete history list
# Environment

**Delta-rs version**:
0.5.4
**Binding**: 
python 3.9.7
**Environment**:
- **Cloud provider**:  IBM
- **OS**: CentOS
- **Other**:

***
# Bug

**What happened**:
The length of the return result of the history function is always between 0 and 9, even if the actual length is greater than 9.

**What you expected to happen**:
return all the history list within 30 days(default `logRetentionDuration`)

**How to reproduce it**:
It is easy to reproduce it. 
1. create a delta table with more than 10 histories.
2. load it with `delta-rs`
3. type history() function

for example:
here is my test data
```shell
(base) ➜  ~ mc ls myhost/mybucket/path/test-delta-rs/_delta_log
[2021-11-26 10:01:50 CST] 1.6KiB 00000000000000000000.json
[2021-11-26 10:02:42 CST] 2.4KiB 00000000000000000001.json
[2021-11-26 10:03:24 CST] 1.1KiB 00000000000000000002.json
...
[2021-12-07 13:51:19 CST]  33KiB 00000000000000000110.checkpoint.parquet
[2021-12-07 13:51:16 CST] 2.1KiB 00000000000000000110.json
[2021-12-07 13:51:23 CST] 2.1KiB 00000000000000000111.json
[2021-12-07 13:51:29 CST] 2.1KiB 00000000000000000112.json
[2021-12-07 13:51:34 CST] 2.1KiB 00000000000000000113.json
[2021-12-07 13:51:39 CST] 2.1KiB 00000000000000000114.json
[2021-12-07 13:51:45 CST] 2.1KiB 00000000000000000115.json
[2021-12-07 13:51:50 CST] 2.1KiB 00000000000000000116.json
[2021-12-07 13:51:55 CST] 2.1KiB 00000000000000000117.json
[2021-12-07 13:52:01 CST] 2.1KiB 00000000000000000118.json
[2021-12-07 13:52:06 CST] 2.1KiB 00000000000000000119.json
[2021-12-07 13:52:14 CST]  35KiB 00000000000000000120.checkpoint.parquet
[2021-12-07 13:52:11 CST] 2.1KiB 00000000000000000120.json
[2021-12-07 13:52:17 CST] 2.1KiB 00000000000000000121.json
[2021-12-07 13:52:23 CST] 2.1KiB 00000000000000000122.json
[2021-12-07 14:15:00 CST] 1.2KiB 00000000000000000123.json
[2021-12-07 14:41:44 CST] 1.3KiB 00000000000000000124.json
[2021-12-07 13:52:14 CST]    27B _last_checkpoint
(base) ➜  ~ 

```

here is my test codes:
```python
>>> from deltalake import DeltaTable
>>> path = 's3://mybucket/path/test-delta-rs'
>>> dt = DeltaTable(path)
>>> dt.version()
124
>>> len(dt.history())
4
>>> dt.load_version(111)
>>> len(dt.history())
1
>>>
```

**More details**:
I think it might be due to `RawDeltaTable` load data from the latest checkpoint and incrementally applying each version since. so the `history()` only returns the incrementally logs from the latest checkpoint.

### PR DESCRIPTION
**Pull request by viirya**
# Description
The description of the main changes of your pull request

# Related Issue(s)

 closes #520 

# Documentation

### CODE REVIEW

File: rust/src/delta.rs [hunk 1]
```diff
@@ -839,14 +839,47 @@ impl DeltaTable {
...
-            None => Ok(commit_infos_list.collect()),
+        let mut version = 0;
+        let mut commit_infos_list = vec![];
+
+        loop {
+            match DeltaTableState::from_commit(self, version).await {
+                Ok(state) => {
+                    commit_infos_list.append(state.commit_infos().clone().as_mut());
+                    version += 1;
+                    println!("version: {:?}", version);
```
Comment (houqp) [inline rust/src/delta.rs:854]: left over debug print? ;)

File: rust/src/delta.rs [hunk 2]
```diff
@@ -839,14 +839,47 @@ impl DeltaTable {
...
-    pub fn history(
+    pub async fn history(
         &mut self,
         limit: Option<usize>,
     ) -> Result<Vec<Map<String, Value>>, DeltaTableError> {
-        let commit_infos_list = self.state.commit_infos().iter().rev().map(Map::clone);
-        match limit {
-            Some(l) => Ok(commit_infos_list.take(l).collect()),
-            None => Ok(commit_infos_list.collect()),
+        let mut version = 0;
```
Comment (houqp) [inline rust/src/delta.rs:846]: The scala reference implementation has an optimization to calculate start probing version using limit and current table version if limit is provided by the user. I think we can adopt the same technique here.

**Review by houqp — COMMENTED** [review]
LGTM overall, left two minor comments :)

File: rust/src/delta.rs (hunk 1 above)
Comment (viirya) [reply rust/src/delta.rs:854]: oops! :)

File: rust/src/delta.rs [hunk 3]
```diff
@@ -839,14 +839,47 @@ impl DeltaTable {
...
+        let mut commit_infos_list = vec![];
+
+        loop {
+            match DeltaTableState::from_commit(self, version).await {
+                Ok(state) => {
+                    commit_infos_list.append(state.commit_infos().clone().as_mut());
+                    version += 1;
+                    println!("version: {:?}", version);
+                    match limit {
+                        Some(history_limit) => {
```
Comment (fvaleye) [inline rust/src/delta.rs:856]: A "if" condition should be sufficient for this, WDYT?

**Review by fvaleye — COMMENTED** [review]
Thanks for the fix @viirya
LGTM! I left only one minor comment

File: rust/src/delta.rs (hunk 3 above)
Comment (viirya) [reply rust/src/delta.rs]: I removed this and added the optimization as Scala implementation.

File: rust/src/delta.rs (hunk 2 above)
Comment (viirya) [reply rust/src/delta.rs]: Good idea. Added it.

File: rust/src/delta.rs [hunk 4]
```diff
@@ -839,14 +869,43 @@ impl DeltaTable {
...
+    pub async fn history(
         &mut self,
         limit: Option<usize>,
     ) -> Result<Vec<Map<String, Value>>, DeltaTableError> {
-        let commit_infos_list = self.state.commit_infos().iter().rev().map(Map::clone);
-        match limit {
-            Some(l) => Ok(commit_infos_list.take(l).collect()),
-            None => Ok(commit_infos_list.collect()),
+        let mut version = match limit {
+            Some(l) => max(self.get_latest_version().await? - l as i64 + 1, 0),
```
Comment (houqp) [inline rust/src/delta.rs:879]: nit, I think using `self.version` here should be good enough so the returned history is consistent with the current table state. If the user wants the latest history, he/she can manually call the update method to catch up.

**Review by houqp — DISMISSED** [review]

File: rust/src/delta.rs (hunk 4 above)
Comment (viirya) [reply rust/src/delta.rs]: Fixed. Thanks.

**Review by houqp — APPROVED** [review]

**Review by fvaleye — APPROVED** [review]

**Comment by houqp** [comment]
Thanks @viirya !

**Comment by viirya** [comment]
Thank you @houqp @fvaleye !

**Pull request merged**